From c59e995c4e719c445e2133b34e7b8e0f71884459 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Tue, 28 Oct 2025 09:04:58 +0800
Subject: [PATCH 4/5] mtk: mt76: mt7996: fix and refactor air monitor

When statin connects to AP, check if station is being
monitored by link address instead of mld address.
Without this patch, mld station cannot be removed from
monitor entry automatically when station is connected.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/main.c   | 2 +-
 mt7996/mt7996.h | 2 +-
 mt7996/vendor.c | 4 ++--
 3 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index a43d4ee96..d9c063920 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1296,7 +1296,7 @@ mt7996_mac_sta_init_link(struct mt7996_dev *dev,
 	ewma_avg_signal_init(&msta_link->avg_ack_signal);
 
 #ifdef CONFIG_MTK_VENDOR
-	mt7996_vendor_amnt_sta_remove(link->phy, sta);
+	mt7996_vendor_amnt_sta_remove(link->phy, link_sta);
 #endif
 
 error:
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 2c0bcc894..a0072185f 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -1518,7 +1518,7 @@ void mt7996_set_wireless_vif(void *data, u8 *mac, struct ieee80211_vif *vif);
 void mt7996_vendor_register(struct mt7996_phy *phy);
 void mt7996_vendor_amnt_fill_rx(struct mt7996_phy *phy, struct sk_buff *skb);
 int mt7996_vendor_amnt_sta_remove(struct mt7996_phy *phy,
-				  struct ieee80211_sta *sta);
+				  struct ieee80211_link_sta *link_sta);
 void mt7996_set_wireless_amsdu(struct ieee80211_hw *hw, u8 en);
 void mt7996_mcu_set_mimo(struct mt7996_phy *phy);
 int mt7996_set_muru_cfg(struct mt7996_dev *dev, u8 action, u8 val);
diff --git a/mt7996/vendor.c b/mt7996/vendor.c
index cb13e7ad7..3a0806bc5 100644
--- a/mt7996/vendor.c
+++ b/mt7996/vendor.c
@@ -558,7 +558,7 @@ mt7996_vendor_amnt_ctrl(struct wiphy *wiphy, struct wireless_dev *wdev,
 }
 
 int mt7996_vendor_amnt_sta_remove(struct mt7996_phy *phy,
-				  struct ieee80211_sta *sta)
+				  struct ieee80211_link_sta *link_sta)
 {
 	u8 zero[ETH_ALEN] = {};
 	int i;
@@ -567,7 +567,7 @@ int mt7996_vendor_amnt_sta_remove(struct mt7996_phy *phy,
 		return 0;
 
 	for (i = 0; i < MT7996_AIR_MONITOR_MAX_ENTRY; i++)
-		if (ether_addr_equal(sta->addr, phy->amnt_ctrl.entry[i].addr))
+		if (ether_addr_equal(link_sta->addr, phy->amnt_ctrl.entry[i].addr))
 			return mt7996_vendor_amnt_set_addr(phy, i, zero);
 	return 0;
 }
-- 
2.45.2

